src/hotspot/share/runtime/thread.hpp

```diff
@@ -806,24 +804,19 @@ class JavaThread: public Thread {
   bool is_trace_suspend()      { return (_suspend_flags & _trace_flag) != 0; }
   bool is_obj_deopt_suspend()  { return (_suspend_flags & _obj_deopt) != 0; }
 
-  // Asynchronous exceptions support
+  // Asynchronous exception support
  private:
-  oop     _pending_async_exception;
-#ifdef ASSERT
-  bool    _is_unsafe_access_error;
-#endif
+  friend class InstallAsyncExceptionHandshake;
+  friend class AsyncExceptionHandshake;
+  friend class HandshakeState;
 
-  inline bool clear_async_exception_condition();
+  void install_async_exception(AsyncExceptionHandshake* aec = NULL);
+  void handle_async_exception(oop java_throwable);
  public:
-  bool has_async_exception_condition() {
-    return (_suspend_flags & _has_async_exception) != 0 &&
-           (_suspend_flags & _async_delivery_disabled) == 0;
-  }
-  inline void set_pending_async_exception(oop e);
+  bool has_async_exception_condition(bool ThreadDeath_only = false);
   inline void set_pending_unsafe_access_error();
   static void send_async_exception(JavaThread* jt, oop java_throwable);
-  void send_thread_stop(oop throwable);
-  void check_and_handle_async_exceptions();
+  static void send_async_exception(JavaThread* jt, AsyncExceptionHandshake* aec);
```

Comment:
I added that overload for Panama, in case they decide to use asynchronous handshakes to solve their cleanup of mapped memory issues. Same thing with virtual method should_throw() in AsyncExceptionHandshake, since they wanted to do some check first in the handshake to decide if they should throw. But I could remove them.

Reply:
Better to not add them until actually needed. Thanks.

Reply:
If you decide to keep it, please add a comment that is there for Panama's use.

Reply:
I removed it. Adding it later if needed sounds good.